Edge growth refactor

chrisbstorey opened the pull request:
Opened on behalf of @isaac-tfn - this superseeds #984

mishtaiwi committed:
Rewrite EDGE replicant in numpy

mishtaiwi committed:
Sectorise Matrices Script

mishtaiwi committed:
Re: #974

- Rename headers of output CSVs
- Sort CSVs to avoid csv>mat errors

mishtaiwi committed:
Check for not grown demand proportion.

isaac-tfn committed:
Small changes for extractor to work

isaac-tfn committed:
Created module for preprocessing growth_matrices
Created module for handling ticket splits
Created module for general utils (mainly matrix manipulations and conversions)
Deleted functions from edge_replicant as they're moved to new modules
Nothing new tested yet

isaac-tfn committed:
Run through to get processed and infilled growth matrices and ticket type splits.
Need to look at caching preprocessing steps.

isaac-tfn committed:
minor change to config yml used for testing

isaac-tfn committed:
Multiprocessing is working - code needs tidying up.

isaac-tfn committed:
Changed multiprocessing so that it doesn't need to write out intermediary variables and read them back in. Small change to config class to make sure export dir exists.

isaac-tfn committed:
Tidied up a bit and added some documentation

isaac-tfn committed:
Created module for loading in global variables and a class for storing them. Altered ticket splits to be able to load directly if available.

isaac-tfn committed:
Ticket type splits are cached when created, and the config can be altered to load this cache in instead of generating splits internally, saving a few minutes

isaac-tfn committed:
Altered run_forecast to run new process instead of edge_replicant for edge. Changed import structure of EDGE_growth.

isaac-tfn committed:
Documenting for a config class

isaac-tfn committed:
Fixed typo in class name

isaac-tfn committed:
Deleted part of run script. Runs from run_forecast now

isaac-tfn committed:
Crude check for existing ticket splits if not provided explicitly

isaac-tfn committed:
Added documentation to config module

isaac-tfn committed:
fji

isaac-tfn committed:
changes for testing

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            userclass = row.Userclass
+            purpose = row.Purpose
+            # read demand matrix
+            zonal_base_demand_mx = utils.wide_mx_2_long_mx(
+                omx_mat.get_matrix_level(segment),
+                rows="from_model_zone_id",
+                cols="to_model_zone_id",
+            )
```
There's some functionality in caf.toolkit to do this - it would be worth digging out everywhere this is used and replacing it with that functionality. Lets not repeat code if we don't have to.

isaac-tfn replied:
Looks like I've been going through comments in the wrong order! Addressed elsewhere and I hopefully will have changed all of these by the time you read this!

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+                )
+                continue
+            # reduce probabilities to current userclass
+            irsj_probs_segment = irsj_props.loc[
+                irsj_props["userclass"] == userclass
+            ].reset_index(drop=True)
```
If we only need prupose segments for each of these iterations, is it worth splitting the irsj file out into multiple? IIRC this is one of the biggest files to read in?

isaac-tfn replied:
Do you mean split the actual file up and read in a new file each loop instead of just taking a subset of the dataframe?

BenTaylor-TfN reviewed (commented):
This review might come out looking a bit weird - its been a long one as I've properly deep dived into it and it's been quite a big overhaul - in future we should probably break this down into chunks.
I'll post this review and come back and finish it off later with more comments

### Overall comments
 - It looks like black isn't picking up the correct settings from the `pyproject.toml` file.
 - There's a couple places where the same issue pops up a few times - I've only commented on the first, but it applies to all of them. I reviewed in the order the code runs.
 - More comments across the code would be better. I expect to see functions split into "blocks" of code, each with a comment explaining _why_ you're doing the things you're doing. It just makes the code easier to visually parse and read.
 - The functions could do with more documentation to make it clearer what each function does and what they return

### Ticket type splits
I know this is where you mentioned there could be an issue so a few ideas:
 - Could you read in the matrices and just ensure that they all sum to one where they should?
 - It looks like one of the purpose keys is `nan`. I imagine this isn't the problem - but could you check why this is happening and solve it?
 - Is there a constant value somewhere that makes sure the ticket type splits, the growth factors, and the translated demand matrices are all in the same order once they're translated into numpy matrices? It looks like there's some implicit ordering, that's easy to fall out of alignment - perhaps we need something more explicit defined as an input file?
   - It could also be worth considering using pandas wide matrices instead of numpy arrays - as long as the index and column names are the zone ids, `.values` will return a numpy array to do all the operations. This means the link between data and zones is retained while keeping the speed improvements of numpy

Looking at the code, I can't see anything obviously wrong that would indicate the wrong splits coming out of here. Have you checked the before code and double checked that there's no bugs in that code too?

isaac-tfn committed:
Small name changes

isaac-tfn committed:
Merge branch 'EDGE_growth_refactor' of https://github.com/Transport-for-the-North/NorMITs-Demand into EDGE_growth_refactor

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,204 @@
+    # fill na
+    edge_flows = edge_flows.fillna(0)
+    # allocate distance bands
+    edge_flows["TAG_Flow"] = edge_flows[
+        edge_flows["TAG_NonDist"].str.startswith("outside south east")
+    ].apply(append_dist, axis=1)
+    edge_flows["TAG_Flow"] = edge_flows["TAG_Flow"].fillna(
+        edge_flows["TAG_NonDist"]
+    )
```
This whole mapping process just got more complicated because I misunderstood it slightly at first and then had to tack bits on yes

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,204 @@
+        ticket_split_proportions, how="left", on=["TAG_Flow"]
+    )
+    # get list of purposes
+    purposes = edge_flows["Purpose"].drop_duplicates().to_list()
```
This was just left alone from Mahmoud's code - I always use unique() but google/stackoverflow seemed to suggest drop_duplicates was identical in this intance. I will change to unique for readability and also probably better to use the same method everywhere.

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/utils.py:
```diff
@@ -0,0 +1,398 @@
+# # # CLASSES # # #
+
+# # # FUNCTIONS # # #
+def long_mx_2_wide_mx(
```
It seems like this is an almost exact duplicate of the toolkit version so I'll switch over to that and delete

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,204 @@
+    ticket_splits_df = file_ops.read_df(
+        ticket_split_params.splits_path
+    )
+    for tp in ["AM", "IP", "PM", "OP"]:
```
Will make them an argument that defaults to what's included.

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,204 @@
+            ticket_splits_df,
+        )
+        split_dict[tp] = splitting_matrices
+    with open(dist_dir / "splitting_matrices.pkl", "wb") as file:
+        pickle.dump(split_dict, file)
+    ticket_split_params.save_yaml(dist_dir / "ticket_split_params.yml")
```
There is a yml file output with some info, do you want a readme too?

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/growth_matrices.py:
```diff
@@ -0,0 +1,163 @@
+    missing_factors: dict
+        dict of missing factors. Gives indices of missing factors.
+    """
+    # order S: R, F
```
I'm not sure what you're referring to here

BenTaylor-TfN replied:
Could you make an enum or something similar that defines the order of priority for each of these types of tickets when replacing growth factors. Then have arguments something like this to achieve the same functionality as atm

```
def fill_missing_factors(
     purposes: list,
     growth_matrices: dict,
     infill_order_full = InfillOrder.FRS,
     infill_order_reduced = InfillOrder.RFS,
     infill_order_season = InfillOrder.SRF,
) -> dict:
```

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/loading.py:
```diff
@@ -0,0 +1,103 @@
+    ]
+    # all segments
+    all_segments = demand_segments["Segment"].to_list()
+    demand_segments.loc[:, "ToHome"] = demand_segments["ToHome"].astype(
```
I started defaulting to .loc to avoid all of the error messages on index vs slicer or whatever it comes up with. Is there a fundamental issue with using .loc or is it just a readbility thing?

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/loading.py:
```diff
@@ -0,0 +1,103 @@
+        with open(params.ticket_type_splits, "rb") as file:
+            ticket_type_splits = pickle.load(file)
+    else:
+        tick_param_path = params.matrices_to_grow_dir / "ticket_split_params.yml"
```
This is sort of just a placeholder as the cache location for this hadn't been decided yet - it was somewhat of a last minute addition. I think this is mentioend elsewhere as well and a lot of file structure/save locations etc. do need overhauling

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/loading.py:
```diff
@@ -0,0 +1,103 @@
+    model_stations_tlcs = file_ops.read_df(
+        params.norms_to_edge_stns_path
+    )
+    if isinstance(params.ticket_type_splits, Path):
+        with open(params.ticket_type_splits, "rb") as file:
+            ticket_type_splits = pickle.load(file)
+    else:
+        tick_param_path = params.matrices_to_grow_dir / "ticket_split_params.yml"
+        if tick_param_path.is_file():
+            ex_params = forecast_cnfg.TicketSplitParams.load_yaml(tick_param_path)
+            if ex_params == params.ticket_type_splits:
+                with open(params.matrices_to_grow_dir / "splitting_matrices.pkl", "rb") as file:
+                    ticket_type_splits = pickle.load(file)
+            else:
+                ticket_type_splits = ticket_splits.splits_loop(
+                    params.ticket_type_splits,
+                    model_stations_tlcs,
+                    params.matrices_to_grow_dir,
+                )
+        else:
+            ticket_type_splits = ticket_splits.splits_loop(
+                params.ticket_type_splits,
+                model_stations_tlcs,
+                params.matrices_to_grow_dir,
+            )
```
Do you mean this whole if/else loop? If so yes it can be. This was also quite last minute and can probably be done more elegantly

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/loading.py:
```diff
@@ -0,0 +1,103 @@
+                model_stations_tlcs,
+                params.matrices_to_grow_dir,
+            )
+    vars = GLobalVars(
```
Good point

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+import sys
+import pickle
+
+sys.path.append(r"E:\NorMITs-Demand")
```
It means it always runs regardless of base directory

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            globals.purposes,
+            growth_matrix_dic,
+        )
+        for i, j in missing_factors.items():
+            j.to_csv(
+                params.export_folder
+                / f"missing_factors_{i}_{forecast_year}.csv"
+            )
```
Possibly could be done somewhere else?

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+                f"{forecast_year}_Demand": [],
+            }
+        )
+        with open(
+            r"E:\NorMITs Demand\Forecasting\edge\1.0\iter1.0\Test\refactor_growth.pkl",
+            "wb",
+        ) as file:
+            pickle.dump(filled_growth_matrices, file)
```
Yes thanks!

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            pickle.dump(filled_growth_matrices, file)
+
+        # loop over time periods
+        tp_looper = partial(
+            _tp_loop,
+            factored_matrices.copy(),
+            params.matrices_to_grow_dir,
+            globals.station_tlcs,
+            forecast_year,
+            globals.demand_segments.copy(),
+            filled_growth_matrices.copy(),
+            growth_summary,
+        )
```
For now I've removed '.copy()'s as I think it's references passed

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            filled_growth_matrices.copy(),
+            growth_summary,
+        )
+        args = [
+            (time_period, globals.ticket_type_splits[time_period])
+            for time_period in globals.time_periods
+        ]
```
Only over multiple lines with incorrect formatter settings!

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            (time_period, globals.ticket_type_splits[time_period])
+            for time_period in globals.time_periods
+        ]
+        outputs = concurrency.multiprocess(
```
I'll put it on my to do list and try and write it once I'm through other comments

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            tp_looper, args, in_order=True
+        )
+        factored = [k["factored"] for k in outputs]
+        factored_matrices = {
+            i: j for i, j in zip(globals.time_periods, factored)
+        }
```
Thanks

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+                f" Percentage of the demand not being grown {round(not_grown_demand_pcent * 100,3)}% is > 1%.\n"
+                "User must review growth factors used."
+            )
+            sys.exit()
```
Changed to ValueError

isaac-tfn committed:
Mid change commit to switch to vm

isaac-tfn commented on normits_demand/models/forecasting/EDGE_growth/utils.py:
```diff
@@ -0,0 +1,398 @@
+    )
+    factors_df.dropna(axis=0, inplace=True)
+    # rename
+    factors_df = factors_df.rename(
+        columns={
+            "stn_zone_id": "to_stn_zone_id",
+            "Demand_rate": "Demand",
+        }
+    )
```
The same columns don't always need to be retained and it seemed easier just to drop outside of the function - all of this process is only done twice so it possibly wasn't the most appropriate pull into a function like this

isaac-tfn committed:
Outputs match previous iteration. Not all comments addressed.

isaac-tfn committed:
All messages resolved. Not all dealt with but committing what is to start working through issues.

isaac-tfn committed:
Created new dataclass for return from _tp_loop

isaac-tfn committed:
Added in two new dirs to config:
ticket_splits_dir is for ticket splits and/or files for creating them
irsj_props_dir is for irsj_probabilty files
These can be set to be equal to matrices_to_grow_dir, as currently all of these files are in that directory

isaac-tfn committed:
Removed 'missing_factors' as outputs were messy and there are multiple other checks for this.

isaac-tfn committed:
small change

isaac-tfn committed:
Fixed some small changes around multiprocessing outputs

isaac-tfn committed:
Removed unused imports

isaac-tfn committed:
Tiny change to a string format

BenTaylor-TfN committed:
Merge branch 'master' into EDGE_growth_refactor

BenTaylor-TfN commented on config/forecast/edge/edge_stn_forecast.yml:
```diff
@@ -4,23 +4,23 @@ export_folder: E:\NorMITs Demand
 
-# EDGE Inputs
-edge_flows_path: I:\NorMITs Demand\import\edge_replicant\edge_inputs\Base and Mappings\FC_fullfat1819_NEW_DDG.csv
+cube_exe: C:\Program Files\Citilabs\CubeVoyager\VOYAGER.EXE
```
Where is this used in the code? Are we actually interfacing with CUBE during this process anymore?

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/__init__.py:
It doesn't look like this update is being picked up by GitHub, perhaps it needs updating directly on here? It's likely because windows is not case-sensitive with its file paths

BenTaylor-TfN commented on normits_demand/models/forecasting/edge_replicant.py:
Is this file still needed at all? It looks like it's the old code.

In my editor, it looks like many of the functions called in here no longer exist

BenTaylor-TfN commented on normits_demand/models/forecasting/forecast_cnfg.py:
```diff
@@ -196,11 +226,63 @@ def export_path(self) -> Path:
+    forecast_years: Self explanatory
+
+    iteration: used in naming dir
+    forecast_version: used in naming dir
```
It won't let me comment on the actual line of code, but I don't like that the `forecast_version` number is hardcoded into this class.

I think hiding away this parameter can make the output path a little confusing too, it confused me when coming back to this code after a while. 

The value likely needs a more descriptive name, such as `EDGE_version`, and be required in the config.

BenTaylor-TfN commented on normits_demand/models/forecasting/forecast_cnfg.py:
```diff
@@ -196,11 +226,63 @@ def export_path(self) -> Path:
+    edge_flows_path: Path
+    flow_cat_path: Path
+    splits_path: Path
+    info: Optional[str]
```
How is the info parameter used? The docstring doesn't explain much

BenTaylor-TfN commented on normits_demand/models/forecasting/forecast_cnfg.py:
```diff
@@ -196,11 +226,63 @@ def export_path(self) -> Path:
+    """Parameters for the EDGE forecasting model.
+    forecast_years: Self explanatory
+
+    iteration: used in naming dir
```
This docstring could probably do with an update to have more useful messages.

Quite a few parameters are also missing here:
- `irsj_props_dir`
- `norms_to_edge_stns_path`
- `cube_exe`
- `edge_growth_dir`

BenTaylor-TfN commented on normits_demand/models/forecasting/forecast_cnfg.py:
```diff
@@ -196,11 +226,63 @@ def export_path(self) -> Path:
+    flow_cat_path: Path
+        Path to flow cat file
+    splits_path: Path
+        Dir containing TicketTypeSplits.csv as well as stn2stn_costs csvs (for each time period). This doubles as the cache
```
Can the path to `TicketTypeSplits.csv` be a separate path please?
The Stn2Stn csvs all come from CUBE, whereas the ticket type splits come from elsewhere and are stored on I drive.

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+            overall_not_grown_demand / overall_base_demand
+        )
+        # if proportion is greater than 1%, terminate the program
+        if not_grown_demand_pcent > 0.01:
```
Would we want to make this an optional parameter in the config?

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+        # empty dataframe for growth summary
+
+        # loop over time periods
+        for tp in global_params.time_periods:
+            hdf_file = params.irsj_props_dir / f"{tp}_iRSj_probabilities_split.h5"
+            if not os.path.isfile(hdf_file):
+                utils.split_irsj(params.irsj_props_dir, 'userclass', tp)
+        tp_looper = partial(
+            _tp_loop,
+            params.matrices_to_grow_dir,
+            params.irsj_props_dir,
+            len(global_params.station_tlcs),
+            forecast_year,
+            global_params.demand_segments,
+            filled_growth_matrices,
+        )
+        args = [
+            (time_period, global_params.ticket_type_splits[time_period])
+            for time_period in global_params.time_periods
+        ]
+        outputs = multiprocess(
+            tp_looper, args, in_order=True
+        )
+        factored = [k.factored for k in outputs]
+        factored_matrices = dict(zip(global_params.time_periods, factored))
+        overall_not_grown_demand = sum([i.overall_not_grown for i in outputs])
+        overall_base_demand = sum([i.overall_base for i in outputs])
```
You don't need the square brackets when calling `sum()`. It is more efficient with generators than lists - much more for massive lists!

BenTaylor-TfN replied:
Is this chunk of code just for converting matrices, growing, then converting back?
Can this be placed into a function to make this more obvious

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+def run_edge_growth(params: forecast_cnfg.EDGEParameters) -> None:
+    """Run Growth Process."""
+    LOG.info("#" * 80)
+    LOG.info(
+        "Started Process @ %s",
+        timing.get_datetime("%d-%m-%Y  %H:%M:%S"),
+    )
+    LOG.info("#" * 80)
```
```suggestion
    LOG.info("Started applying EDGE Growth")
    LOG.info("#" * 80)
```

We don't need the datetime being stored as it is part of all the log messages anyway

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+    global_params = loading.load_globals(params)
+    # loop over forecast years
+    for forecast_year in params.forecast_years:
+        LOG.info(
+            "**** Applying growth for %s @ %s",
+            forecast_year,
+            timing.get_datetime(),
+        )
```
```suggestion
        LOG.info("Applying growth for %s", forecast_year)
```

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/__init__.py:
```diff
@@ -0,0 +1,3 @@
+import sys
+
+from normits_demand.models.forecasting.edge_growth.run import run_edge_growth
```
```suggestion
from normits_demand.models.forecasting.edge_growth.run import run_edge_growth
```

`sys` isn't used!

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+    )
+    LOG.info("#" * 80)
+    LOG.info("Loading global variables for growth process.")
+    global_params = loading.load_globals(params)
```
I'd avoid calling anything global in code - global already has connotations, which can be misleading in this context

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+        # fill growth matrices
+        filled_growth_matrices= growth_matrices.fill_missing_factors(global_params.purposes,growth_matrix_dic)
+        # create empty dictionary to store matrices
+        factored_24hr_matrices = {}
```
Move this closer to where the variable is actually used

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+        filled_growth_matrices= growth_matrices.fill_missing_factors(global_params.purposes,growth_matrix_dic)
+        # create empty dictionary to store matrices
+        factored_24hr_matrices = {}
+        # declare global demand total variables
+        # empty dataframe for growth summary
```
```suggestion
```

Are these stray comments?

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+        # declare global demand total variables
+        # empty dataframe for growth summary
+
+        # loop over time periods
```
Useless comment - what is this loop actually doing?

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+# # # CLASSES # # #
+
+# # # FUNCTIONS # # #
+def _tp_loop(
```
I know this is only an internal function, but a more descriptive name would be helpful for review. It's not clear at the outset _why_ we are looping over time periods.

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,342 @@
+            (
+                np_stn2stn_base_demand_mx,
+                zonal_from_to_stns,
+            ) = utils.zonal_from_to_stations_demand(
```
Tacking onto this comment - it looks like the "sister" function is named `convert_stns_to_zonal_demand()`. Could you convert this name to `convert_zonal_to_stns_demand()` to make the relationship between the two clearer?

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+            f"PT_{time_period}.omx",
+        )
+    ) as omx_mat:
+        for row in tqdm(
+            demand_segments.itertuples(),
+            desc="    Demand Segments Loop ",
+            unit=" Segment",
+            colour="cyan",
+            total=len(demand_segments),
+        ):
```
```suggestion
        for row in tqdm(
            demand_segments.itertuples(),
            desc="Growing matrices",
            unit="Segment",
            colour="cyan",
            total=len(demand_segments),
        ):
```

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+        f"{'Base_Demand':>12}{f'{forecast_year}_Demand':>12}"
+    )
+
+    with omx_file.OMXFile(
+        pathlib.Path(
+            matrices_to_grow_dir,
+            f"PT_{time_period}.omx",
+        )
+    ) as omx_mat:
```
```suggestion
    file_path = pathlib.Path(matrices_to_grow_dir, f"PT_{time_period}.omx")
    with omx_file.OMXFile(file_path) as omx_mat:
```

Much easier to read

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/run.py:
```diff
@@ -0,0 +1,315 @@
+                    f"{forecast_year}_Demand": [tot_output_demand],
+                }
+            )
+            growth_summary = pd.concat(
+                [growth_summary, segment_growth_summary], axis=0
+            )
```
Do we need this DataFrame inside the loop at all? 

If not, its probably quicker just to store a list of dictionaries, then convert that into a DataFrame at the end of the loop.

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/growth_matrices.py:
```diff
@@ -0,0 +1,149 @@
+
+    # get growth matrices for each purpose/ticket type
+    for purpose in purposes:
+        for ticket_type in ticket_types:
+            mx_df = factors_df[
+                ["from_stn_zone_id", "to_stn_zone_id", "Demand"]
+            ].loc[
+                (factors_df["purpose"] == purpose)
+                & (factors_df["TicketType"] == ticket_type)
+            ]
+            # expand matrix
+            index = range(1, len(stations_lookup)+1)
+            growth_matrices[purpose][
+                ticket_type
+            ] = ctk.pandas_utils.long_to_wide_infill(mx_df, mx_df.columns[0], mx_df.columns[1], mx_df.columns[2], index, index, infill=0).values
```
It's more efficient to build a separate dictionary for each ticket type inside the purpose loop, then assign that built dictionary to the purpose-keyed dictionary.

i.e. Rather than
```
growth_matrices = {i: {} for i in purposes}
for purpose in purposes:
    for ticket_type in ticket_types:
        growth_matrices[purpose][ticket_type] = some_operation()
```

Do this:
```
growth_matrices = dict()
for purpose in purposes:
    p_dict = dict()
    for ticket_type in ticket_types:
        p_dict[ticket_type] = some_operation()
    growth_matrices[purpose] = p_dict
```

In the second method you're having to do significantly less hashes to assign values to the dictionary

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/loading.py:
```diff
@@ -0,0 +1,129 @@
+    """
+
+@dataclasses.dataclass
+class GlobalVars:
```
I'd avoid calling anything global in code - global already has connotations, which can be misleading in this context.

i.e. calling `globals()` in code will return all current global variables

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,250 @@
+# # # CLASSES # # #
+
+# # # FUNCTIONS # # #
+def append_dist(row):
```
It would be good for this to be converted into some kind of input file so it's more clear how the TAG flows are being categorised

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/ticket_splits.py:
```diff
@@ -0,0 +1,250 @@
+    else:
+        return row['TAG_NonDist'].lower() + " 100 + miles - adjusted"
+    
+def add_distance_band_tag_flow(mx_df: pd.DataFrame) -> pd.DataFrame:
```
Same as above, this looks like it would be much easier if it was just an input file doing the mapping!

BenTaylor-TfN commented on normits_demand/models/forecasting/EDGE_growth/apply_growth.py:
```diff
@@ -0,0 +1,114 @@
+    return stn2stn_forecast_mx
+
+
+def fromto_2_from_by_averaging(
```
This is a horrible function name! Isn't the averaging only applied to NHB matrices?

BenTaylor-TfN reviewed (commented):
Couple of general comments below:

- I can't quite figure out what's going on during the run, but it looks like there's 2 loggers capturing the `LOG` calls and writing them to the terminal in two different formats. This probably wants tidying up before pushing to master.
- Has `black` been run on the new files? It might help sort out some of the weird formatting going on.
- I know some of this is old code, but there are a lot of comments on a line by line basis, and this is generally bad practice.
- Again, old code, but use of whitespace is very useful when reading code back! I've added some in during the review.

BenTaylor-TfN committed:
Review edits

BenTaylor-TfN committed:
Update norms_stn_forecast.yml

isaac-tfn committed:
Change to netm forecast yml

isaac-tfn committed:
Merge branch 'EDGE_growth_refactor' of https://github.com/Transport-for-the-North/NorMITs-Demand into EDGE_growth_refactor

isaac-tfn committed:
Fixed issue with toolkit imports

isaac-tfn committed:
Getting edge forecasting working

isaac-tfn committed:
Added warning for merge fail in tickettypesplits

isaac-tfn committed:
applicator code running

isaac-tfn committed:
Function to apply covid adjustments

isaac-tfn committed:
Add cap year factor

isaac-tfn committed:
Fix out_year in covid script

isaac-tfn committed:
Recently ran edge adjustments
